kernel/arch/dreamcast/include/dc/maple/controller.h

```diff
@@ -24,149 +27,466 @@
+    this structure.
 
-    /** \brief  Secondary joystick y-axis value (if applicable). */
-    int joy2y;
-} cont_state_t;
+    A 1 bit in the buttons' bitfield indicates that a button is pressed, and the
+    joyx, joyy, joy2x, joy2 values are all 0 based (0 is centered).
 
-/* \cond */
-/* Init / Shutdown */
-int cont_init(void);
-void cont_shutdown(void);
-/* \endcond */
+    \sa maple_dev_status
+*/
+typedef struct cont_state {
+    union {
+        /** \brief  bit-packed controller button states
+            \sa     controller_buttons
+        */
+        uint32_t buttons;
+        struct {
+            uint32_t c: 1;              /**< \brief C button value. */
+            uint32_t b: 1;              /**< \brief B button value. */
+            uint32_t a: 1;              /**< \brief A button value. */
+            uint32_t start: 1;          /**< \brief Start button value. */
+            uint32_t dpad_up: 1;        /**< \brief Main Dpad Up button value. */
+            uint32_t dpad_down: 1;      /**< \brief Main Dpad Down button value. */
+            uint32_t dpad_left: 1;      /**< \brief Main Dpad Left button value. */
+            uint32_t dpad_right: 1;     /**< \brief Main Dpad Right button value. */
+            uint32_t z: 1;              /**< \brief Z button value. */
+            uint32_t y: 1;              /**< \brief Y button value. */
+            uint32_t x: 1;              /**< \brief X button value. */
+            uint32_t d: 1;              /**< \brief D button value. */
+            uint32_t dpad2_up: 1;       /**< \brief Secondary Dpad Up button value. */
+            uint32_t dpad2_down: 1;     /**< \brief Secondary Dpad Down button value. */
+            uint32_t dpad2_left: 1;     /**< \brief Secondary Dpad Left button value. */
+            uint32_t dpad2_right: 1;    /**< \brief Secondary Dpad Right button value. */
+            uint32_t: 16;
+        };
+    };
+
+    int ltrig;    /**< \brief Left trigger value (0-255). */
+    int rtrig;    /**< \brief Right trigger value (0-255). */
+    int joyx;     /**< \brief Main joystick x-axis value. (-128 - 127) */
+    int joyy;     /**< \brief Main joystick y-axis value. */
+    int joy2x;    /**< \brief Secondary joystick x-axis value (if applicable). */
+    int joy2y;    /**< \brief Secondary joystick y-axis value (if applicable). */
+} cont_state_t;
 
-/** \brief  Controller automatic callback type.
+/** \brief   Controller automatic callback type.
+    \ingroup controller_inputs
 
     Functions of this type can be set with cont_btn_callback() to respond
     automatically to the specified set of buttons being pressed. This can be
     used, for instance, to implement the standard A+B+X+Y+Start method of ending
     the program running.
+
+    \warning
+    Your callback will be invoked within a context with interrupts disabled.
+    See cont_btn_callback for more information.
+
+    \param addr             Maple BUS address to poll for the button mask
```

Comment:
Would it be worthwhile to note addr and btns with `maple_addr()` and the defgroup for btns?

Reply:
Oh, hah, you do it with the callback installer below. Excellent!